### PR TITLE
feat: added variants implementation and unit tests

### PR DESCRIPTION
**Pull request by vkalta**

### CODE REVIEW

File: Cartfile [hunk 1]
```diff
@@ -1,2 +1,2 @@
-github "AliSoftware/OHHTTPStubs" "8.0.0"
-github "venmo/DVR" "v2.0.1"
+github "venmo/DVR"
```
Comment (sunil-lakshman) [inline Cartfile:1]: @vkalta Is it required this repo?

File: Cartfile.resolved [hunk 2]
```diff
@@ -1,2 +1,2 @@
-github "AliSoftware/OHHTTPStubs" "8.0.0"
-github "venmo/DVR" "v2.0.1"
+github "vkalta/DVR" "8bb79349473d1ff595f07d216662ee5163854f42"
```
Comment (sunil-lakshman) [inline Cartfile.resolved:1]: @vkalta Please update the repo name and ensure no secrets are included in the PR.

File: Cartfile.resolved [hunk 3]
```diff
@@ -1,2 +1,2 @@
...
+github "vkalta/contentstack-utils-swift" "dacaf16bd8e4e182e0bef57df2957c7c8697082c"
```
Comment (sunil-lakshman) [inline Cartfile.resolved:2]: @vkalta Please update the repo name and ensure no secrets are included in the PR.

File: Contentstack.xcworkspace/xcshareddata/xcschemes/Contentstack macOS Tests.xcscheme [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      selectedDebuggerIdentifier = "Xcode.DebuggerFoundation.Debugger.LLDB"
+      selectedLauncherIdentifier = "Xcode.DebuggerFoundation.Launcher.LLDB"
+      shouldUseLaunchSchemeArgsEnv = "YES"
+      shouldAutocreateTestPlan = "YES">
+      <Testables>
+         <TestableReference
+            skipped = "NO">
+            <BuildableReference
+               BuildableIdentifier = "primary"
+               BlueprintIdentifier = "0F4A75F2241BAE6C00E3A024"
```
Comment (sunil-lakshman) [inline Contentstack.xcworkspace/xcshareddata/xcschemes/Contentstack macOS Tests.xcscheme:21]: @vkalta  Please avoid pushing secrets in the PR.

File: Tests/AssetTest.swift [hunk 5]
```diff
@@ -15,13 +15,13 @@ class AssetTest: XCTestCase {
...
     #if !NO_FATAL_TEST
 //    The test runner exited with code 6 before finishing running tests. This may be due to your code calling 'exit', consider adding a symbolic breakpoint on 'exit' to debug.
-    func testFetch_withoutUID() async {
-        expectFatalError(expectedMessage: "Please provide Asset uid") {
-            makeAssetSut().fetch { (result: Result<AssetModel, Error>, response) in
-                        
-            }
-        }
-    }
+//    func testFetch_withoutUID() async {
```
Comment (sunil-lakshman) [inline Tests/AssetTest.swift:18]: @vkalta  Could you clarify why this test case has been commented out?

File: Tests/ContentTypeTest.swift [hunk 6]
```diff
@@ -16,13 +16,13 @@ final class ContentTypeTests: XCTestCase {
...
     }
     #if !NO_FATAL_TEST
-    func testFetch_withoutUID() async {
-        expectFatalError(expectedMessage: "Please provide ContentType uid") {
-            makeContentTypeSut().fetch { (result: Result<AssetModel, Error>, response) in
-                        
-            }
-        }
-    }
+//    func testFetch_withoutUID() async {
```
Comment (sunil-lakshman) [inline Tests/ContentTypeTest.swift:19]: @vkalta Could you clarify why this test case has been commented out?

File: Tests/ContentstackTests.swift [hunk 7]
```diff
@@ -49,11 +49,6 @@ final class ContentstackTests: XCTestCase {
         XCTAssertEqual(stack.host, host)
     }
 
-    func testStack_Branch_ReturnStackWithBranch () {
```
Comment (sunil-lakshman) [inline Tests/ContentstackTests.swift:52]: @vkalta Could you clarify why this test case code has been deleted?

File: Tests/EntryTest.swift [hunk 8]
```diff
@@ -14,13 +14,13 @@ class EntryTest: XCTestCase {
...
     }
     #if !NO_FATAL_TEST
-    func testFetch_withoutUID() async {
-        expectFatalError(expectedMessage: "Please provide Entry uid") {
-            makeEntrySut(contentTypeuid: "content").fetch { (result: Result<AssetModel, Error>, response) in
-                        
-            }
-        }
-    }
+//    func testFetch_withoutUID() async {
```
Comment (sunil-lakshman) [inline Tests/EntryTest.swift:17]: @vkalta Could you clarify why this test case has been commented out?

File: Tests/SyncTest.swift [hunk 9]
```diff
@@ -32,12 +32,12 @@ class SyncTest: XCTestCase {
         XCTAssertEqual(syncStack.parameter.query(), "pagination_token=\(paginationToken)")
     }
     #if !NO_FATAL_TEST
-    func testSync_BothTokens_ShouldGetFatalError() {
-        expectFatalError(expectedMessage: ("Both Sync Token and Pagination Token can not be presnet.")) {
-            let syncStack = makeSyncStack(syncToken: self.syncToken, paginationToken: self.paginationToken)
-            XCTAssertNil(syncStack)
-        }
-    }
+//    func testSync_BothTokens_ShouldGetFatalError() {
```
Comment (sunil-lakshman) [inline Tests/SyncTest.swift:35]: @vkalta Could you clarify why this test case has been commented out?

File: Cartfile (hunk 1 above)
Comment (vkalta) [reply Cartfile:1]: Hi @sunil-lakshman , yes its required.

File: Cartfile.resolved (hunk 2 above)
Comment (vkalta) [reply Cartfile.resolved:1]: This is not a secret @sunil-lakshman , this is commit hash to ensure the carthage installs from that specific repo.

File: Cartfile.resolved (hunk 3 above)
Comment (vkalta) [reply Cartfile.resolved:2]: This is not a secret @sunil-lakshman, this is commit hash to ensure the carthage installs from that specific commit.
The repos used are updated with the target platform version which is iOS 12. Previously it was 11 and was causing the build to break.

File: Contentstack.xcworkspace/xcshareddata/xcschemes/Contentstack macOS Tests.xcscheme (hunk 4 above)
Comment (vkalta) [reply Contentstack.xcworkspace/xcshareddata/xcschemes/Contentstack macOS Tests.xcscheme:21]: Hi @sunil-lakshman , this is not a secret. This is autogenerated by Xcode to identify schemes.

File: Tests/AssetTest.swift (hunk 5 above)
Comment (vkalta) [reply Tests/AssetTest.swift:18]: I already had mentioned this earlier @sunil-lakshman . This test was causing thread based issues, due to which I had to remove this. This needs to be handled separately or can be ignored for now.

File: Tests/ContentTypeTest.swift (hunk 6 above)
Comment (vkalta) [reply Tests/ContentTypeTest.swift:19]: I already had mentioned this earlier @sunil-lakshman . This test was causing thread based issues, due to which I had to remove this. This needs to be handled separately or can be ignored for now.

File: Cartfile.resolved (hunk 2 above)
Comment (sunil-lakshman) [reply Cartfile.resolved:1]: @vkalta Using your personal account repository **vkalta/DVR** is not allowed by SRE.

File: Cartfile.resolved (hunk 3 above)
Comment (sunil-lakshman) [inline Cartfile.resolved:2]: @vkalta Using your personal account repository **vkalta/contentstack-utils-swift** is not allowed by SRE.

**Review by sunil-lakshman — APPROVED** [review]
Approved

**Pull request merged**